### PR TITLE
Add events to contract suite

### PR DESCRIPTION
**Pull request by sisyphusSmiling**
Closes: #15

## Description

Adding events to HybridCustody, CapabilityProxy, and CapabilityFilter contracts. The initial form of this PR should not be taken as prescriptive or final, but as a concrete starting point to drive discussions around the topic and quickly iterate on if needed.

At a high level, we want a user to know:
1. When they have an account available to claim
2. When they have successfully claimed said account access
3. Identifying info about the resources involved in their access on an account
4. When their account has been removed, either by their direct action or by another party with the ability to do so
5. When a Capability has been added/removed to/from a relevant CapabilityProxy
6. When a new Type has been added/removed to/from a relevant CapabilityFilter
7. When they have been given ownership over an account
8. When an account has been sealed

## Included events

With those questions in mind, here's an overview of the events included in the PR as submitted:

### HybridCustody
```js
pub event CreatedManager(id: UInt64)
pub event CreatedChildAccount(id: UInt64, child: Address)
pub event AddedProxyAccount(id: UInt64, child: Address, parent: Address)
pub event AddedOwnedAccount(id: UInt64, child: Address, parent: Address)
pub event RemovedProxyAccount(id: UInt64?, child: Address, parent: Address)
pub event RemovedOwnedAccount(id: UInt64?, child: Address, parent: Address)
pub event ProxyAccountPublished(
    childAcctID: UInt64,
    proxyAcctID: UInt64,
    capProxyID: UInt64,
    factoryID: UInt64,
    filterID: UInt64,
    filterType: Type,
    child: Address,
    pendingParent: Address
)
pub event ChildAccountRedeemed(id: UInt64, child: Address, parent: Address)
pub event RemovedParent(id: UInt64, child: Address, parent: Address)
pub event OwnershipGranted(id: UInt64, child: Address, parent: Address)
pub event SealedAccount(id: UInt64, address: Address, parents: [Address])
```

### CapabilityProxy
```js
pub event ProxyCreated(id: UInt64)
pub event CapabilityAdded(id: UInt64, type: Type, isPublic: Bool)
pub event CapabilityRemoved(id: UInt64, type: Type)
```

### CapabilityFilter
```js
pub event TypeAdded(id: UInt64, filterType: Type, newType: Type)
pub event TypeRemoved(id: UInt64, filterType: Type, removedType: Type)
```

## Open questions:
- Is it more useful to emit `Type.identifier` or Type in events?
- How much data (if any) do we want to emit about accessible Types in the `ProxyAccountPublished` event? This might be informed by conversations around #18 and #19

### CODE REVIEW

**Comment by sisyphusSmiling** [comment]
Really helpful context and feedback @bjartek, thank you!

What do you think about:

```cadence
ProxyAccountUpdated(id: UInt64?, child: Address, parent: Address, added: Bool)
```

Where:

- `added == true` -> `AddedProxyAccount`
- `added == false` -> `RemovedProxyAccount`

If helpful/necessary, maybe we could extend this further to condense proxy & owned events with an `isOwned: Bool` flag. More specifically this would look like:

```cadence
AccountUpdated(id: UInt64?, child: Address, parent: Address, added: Bool, isOwned: Bool)
```

| Equivalency | `added` | `!added` |
| --- | --- | --- |
| `isOwned` | `AddedOwnedAccount` | `RemovedOwnedAccount` |
| `!isOwned` | `AddedProxyAccount` | `RemovedProxyAccount` |

What do you think?

**Review by austinkline — COMMENTED** [review]
Nice work! First round of comments, didn't want to delay getting my thoughts out.

File: contracts/CapabilityProxy.cdc
Comment (bjartek) [reply contracts/CapabilityProxy.cdc:22]: I second this, but it is not very important for me. Flowscan show types very verbose m. Just check a storefront listing.

File: contracts/HybridCustody.cdc
Comment (sisyphusSmiling) [reply contracts/HybridCustody.cdc:649]: Makes sense. It's more consistent with `AccountUpdated`

File: contracts/CapabilityProxy.cdc
Comment (austinkline) [reply contracts/CapabilityProxy.cdc]: Identifier is more friendly, but it is also more prescriptive in that it prevents anyone who would want to listen to the type from doing so. On the other hand, emitting the event lets anyone who wants to grab the identifier get to it and just replace the field. Because of that, I'd stick with type just so that we don't prevent folks from doing what they want, here. Thoughts?

Comment (sisyphusSmiling) [reply contracts/CapabilityProxy.cdc]: Agreed. Better to have a flexible solution

File: contracts/CapabilityProxy.cdc
Comment (austinkline) [inline contracts/CapabilityProxy.cdc:100]: Leaving this suggestion in case we want to handle the case that a proxy capability exists in both the private and public buckets

```suggestion
            if let removedPublic = self.publicCapabilities.remove(key: cap.getType()) {
                emit ProxyUpdated(id: self.uuid, capabilityType: cap.getType(), isPublic: true, active: false)
            }
            
            if let removedPrivate = self.privateCapabilities.remove(key: cap.getType()) {
                emit ProxyUpdated(id: self.uuid, capabilityType: cap.getType(), isPublic: false, active: false)
            }
```

**Review by rrrkren — APPROVED** [review]
this looks good from dapper wallet perspective. if you can provide a rundown of how the events are expected to be aggregated for a NFT Ownership backend service that syncs with chain through events that'd be great

**Comment by sisyphusSmiling** [comment]
Thanks for the feedback @rrrkren 

> provide a rundown of how the events are expected to be aggregated for a NFT Ownership backend service that syncs with chain through events

Great recommendation. I'll note down how I've thought about this here, but further guidance will be included in dedicated docs.

As of today, aggregators would follow NFT-related `Deposit` and `Withdraw` events, deducing token transfers from `Withdraw.from` to `Deposit.to`.

With the introduction of the HybridCustody contract, there's now a meta-level where account delegation can be transferred and removed. In the context of tracing NFT ownership, this means a parent account has transitive ownership of an NFT owned by a child account.

Take the following example:

`MomentMinted(momentID: 42, playID: 3, setID: 7, serialNumber: 13, subeditionID: 17)`
`TopShot.Deposit(id: 42, to: 0x03)`

From these events, we know `TopShot.NFT` with id 42 was minted and transferred to `0x03`, thus making them the owner. Now, add in the following HybridCustody event:

`AccountUpdated(id: 19, child: 0x03, parent: 0x07, proxy: true, active: true)`

We now have a picture where `0x07` accesses `0x03` which owns `TopShot.NFT` with id 42.

The missing piece here is whether the `ProxyAccount` Capability maintained by `0x07` has the ability to withdraw from the TopShot Collection in `0x03`. In most cases, it's likely true that the parent account would have that access, but not guaranteed. 

An application leveraging HybridCustody constructs to facilitate cross-account interactions would need to query the parent account to understand the access a said account has on a particular child account at a granular level. This issue is covered by the broader topic of "discoverability", and the details of that query and relevant helper methods will be addressed in #19.

**Comment by austinkline** [comment]
@rrrkren and @bjartek anything else on this PR we'd like to address? If not, I think we can let it go. @sisyphusSmiling anything else on your mind before we merge?

**Review by austinkline — APPROVED** [review]

**Comment by sisyphusSmiling** [comment]
@austinkline nothing else to add on my end!

**Comment by austinkline** [comment]
> @austinkline nothing else to add on my end!

Going to merge this in, if anyone else has concerns let's address it in another PR. Nice work!

**Pull request merged**
